### ISSUE TITLE
Warnings about Lock Directory Portability

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#9406]
Since we now record the value of all the variables evaluated during solving, we can now detect if one of these variables does not match the build time environment. For example, if we used `%{os}% = "linux"` to compute some dependencies, we can warn the user trying to run the lock directory on macos.

**Comment by gridbugs** [comment]
Yep, for each combination of os, arch, distro, etc that affects the dependency solution there will be a separate lockdir. 

I'm not sure we should even attempt to proceed with a build if the recorded solver env differs from the current env, at least not by default. We attempt to filter build/install commands at solve time, and if a necessary command for your env was filtered out at solve time it won't be in the lockdir at all and you'll have no way of knowing that this was the case without manually checking the original opam file. This may lead to strange errors which are hard to debug. Since reliable builds is one of the selling points of lockdirs I think dune should be more conservative about attempting builds where this is a possibility.
